examples/pydantic_ai_examples/flight_booking.py

```diff
@@ -0,0 +1,239 @@
+        'Row 1 is the front row and has extra leg room. '
+        'Rows 14, and 20 also have extra leg room. '
+    ),
+)
+
+
+# in reality this would be downloaded from a booking site,
+# potentially using another agent to navigate the site
+flights_web_page = """
+1. Flight SFO-AK123
+- Price: $350
+- Origin: San Francisco International Airport (SFO)
+- Destination: Ted Stevens Anchorage International Airport (ANC)
+- Date: January 10, 2025
+
+2. Flight SFO-AK456
+- Price: $370
+- Origin: San Francisco International Airport (SFO)
+- Destination: Fairbanks International Airport (FAI)
+- Date: January 10, 2025
+
+3. Flight SFO-AK789
+- Price: $400
+- Origin: San Francisco International Airport (SFO)
+- Destination: Juneau International Airport (JNU)
+- Date: January 20, 2025
+
+4. Flight NYC-LA101
+- Price: $250
+- Origin: San Francisco International Airport (SFO)
+- Destination: Ted Stevens Anchorage International Airport (ANC)
+- Date: January 10, 2025
+
+5. Flight CHI-MIA202
+- Price: $200
+- Origin: Chicago O'Hare International Airport (ORD)
+- Destination: Miami International Airport (MIA)
+- Date: January 12, 2025
+
+6. Flight BOS-SEA303
+- Price: $120
+- Origin: Boston Logan International Airport (BOS)
+- Destination: Ted Stevens Anchorage International Airport (ANC)
+- Date: January 12, 2025
+
+7. Flight DFW-DEN404
+- Price: $150
+- Origin: Dallas/Fort Worth International Airport (DFW)
+- Destination: Denver International Airport (DEN)
+- Date: January 10, 2025
+
+8. Flight ATL-HOU505
+- Price: $180
+- Origin: Hartsfield-Jackson Atlanta International Airport (ATL)
+- Destination: George Bush Intercontinental Airport (IAH)
+- Date: January 10, 2025
+"""
+
+# restrict how many requests this app can make to the LLM
+usage_limit = UsageLimits(request_limit=15)
+
+
+async def main():
+    deps = Deps(
+        web_page_text=flights_web_page,
+        req_origin='SFO',
+        req_destination='ANC',
+        req_date=datetime.date(2025, 1, 10),
+    )
+    message_history: list[ModelMessage] | None = None
+    usage: Usage = Usage()
+    while True:
+        result = await search_agent.run(
+            'Find me a flight',
+            deps=deps,
+            usage=usage,
+            message_history=message_history,
+            usage_limits=usage_limit,
+        )
+        if isinstance(result.data, NoFlightFound):
+            print('No flight found')
+            break
+        else:
+            flight = result.data
+            print(f'Flight found: {flight}')
+            answer = Prompt.ask(
+                'Do you want to buy this flight, or keep searching? (buy/*search)',
+                choices=['buy', 'search', ''],
+                show_choices=False,
+            )
+            if answer == 'buy':
+                seat = await find_seat(usage)
+                await buy_tickets(flight, seat)
+                break
+            else:
+                result.set_result_tool_return('Please suggest another flight')
```

Comment:
Is this documented somewhere? I didn't see documentation other than in the docstring, and that wasn't enough for me to understand what the purpose of this `set_result_tool_return` api is. I feel like it might be worth explaining somewhere in the docs (doesn't need to be in this PR), it feels like it could be made more concrete, in particular, if it's updating one of the messages in the history or something.

Reply:
Yes, agreed

Reply:
is the API wrong, or just not documented?

Reply:
If I understand what is happening correctly, I feel like this is a niche enough scenario to not merit a separate API. In particular, it feels very much like it's exposing implementation details to allow you to set the content of the message that was parsed into the result type. Or at least, it feels to me like it adds an API that is only really usable if you fully understand the implementation details (I mean I already feel like I struggle to understand what it does, so I imagine people less familiar with the library will struggle more.)

You could imagine having a validator for the result tool that guarantees something, which could then be changed by this API, which feels unfortunate.

I feel like it makes more sense and generalizes better to, instead of modifying the result message, just add a new message to the message history, i.e., after the line `message_history = result.all_messages()` adding `message_history.append(ModelRequest(parts=[UserPromptPart(content='Please suggest another flight')]))`. (I don't believe we force the requests and responses to be paired, so I think this should be okay?)

If you have a problem with having two request messages rather than one response and one request, I still think it makes more sense to make it so the analogous API modifies the next request, rather than modifying the previous response. (E.g., in the next call to `run`, we could modify the latest item if it is a request, so appending a new ModelRequest would not result in two consecutive requests. I don't think that's necessary though, or even if something is I'm not necessarily convinced it's the best solution.)

Reply:
The point is, you have to have a response to each tool call.

I think what I'm doing in this example will be a very common workflow.

I'm happy to consider another API, but I think being able to customise the return valid of the result tool is important and special enough to get it's own meth.d

Reply:
fixed.